app-experimental.tex

```diff
@@ -2395,3 +2395,219 @@ \section{Distributed Capabilities For Peripherals And Accelerators} % <<<
+
+\subsection{Compartment Change}
+\label{subsec:comp_change}
+
+A compartment change is done when the CID is changed.
+The CID can be changed by installing a code capability with a different CID into the PCC register.
+Installing a new PCC can be facilitated in many ways, e.g., by jumping to a capability using the CJALR instruction.
+We envision there to exist multiple ways of changing the PCC and the CID sealing mechanism is independent of the concrete way.
+Once the new PCC is installed, the new compartment needs to bootstrap.
+We envision two ways for that, which only differ in the way they retrieve their initial data capability.
+
+First, the new PCC can have further capabilities in its global space, which can be loaded into the register file.
+This can be facilitated by the AUIPCC instruction.
+The following instruction sequence loads a capability at an example fixed offset:
+
+\texttt{auipcc ct0, 2}\\
+\texttt{clc cs0, 0x100(ct0)}
+
+Second, a new data capability can be installed alongside the new PCC.
+This gives the compartment a capability ready to use in the register file.
+One option for this kind of compartment transition would be indirect sealed capability pairs.
+The following code example shows the brevity of this approach.
+The instruction sequence was shrunk to just one instruction reading from the new data capability(ct6).
+
+\texttt{clc cs0, 0(ct6)}
+
+\subsection{Sharing}
+
+Sealing explicitly forbids sharing capabilities between compartments.
+However, sometimes this behaviour is desired by software. We have designed two mechanisms for sharing capabilities between compartments:
+
+\begin{itemize}
+\item Explicit unsealing: This will explicitly unseal a capability and it can be shared either via the register file or via memory.
+Another compartment can pick up this capability and seal it again and use it.
+This does not protect against transitive capability leaks - an unsealed capability can be passed on to a third compartment and thus be leaked.
+\item CID spaces: CIDs can be separated into CID spaces (see Section~\ref{subsec:cid_spaces}).
+One approach is to allow unsealing within a CID space.
+This allows all compartments in that space to use that capability, but no other compartment can unseal the capability.
+This protects from transitive capability leaks outside of the CID space.
+As an addition, we envision a bit that -- if set -- allows sharing within the CID space.
+\end{itemize}
+
+Furthermore, we also envision the possibility of \textit{resealing} capabilities.
+A capability belonging to one compartment can be resealed to another compartment.
+This effectively means that code can seal one or more of its own capabilities for another compartment.
+The main advantages are that this avoids capabilities to be unsealed in the open as done with explicit unsealing as well as potential performance improvements due not needing to unseal and seal capabilites.
```

Comment:
> potential performance improvements due not needing to unseal and seal capabilites.
-> "potential performance improvements due *to* not needing to unseal and seal capabilites." or "potential performance improvements *avoiding the need* to unseal and seal capabilites."